src/main.rs

```diff
@@ -1,3 +1,192 @@
+                    .headers()
+                    .get("X-Influxdb-Version")
+                    .unwrap()
+                    .to_str()
+                    .unwrap();
+
+                (String::from(build), String::from(version))
+            })
+            .map_err(|err| println!("request error: {}", err))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{InfluxDbClient, InfluxDbQuery};
+    use tokio::runtime::current_thread::Runtime;
+
+    fn get_runtime() -> Runtime {
+        Runtime::new().expect("Unable to create a runtime")
+    }
+
+    fn create_client() -> InfluxDbClient {
+        InfluxDbClient {
+            _url: String::from("http://localhost:8086"),
+            _database: String::from("test"),
+        }
+    }
+
+    #[test]
+    fn test_ping() {
+        let client = create_client();
+        let result = get_runtime().block_on(client.ping());
+        assert!(!result.is_err(), "Should be no error");
+
+        let (build, version) = result.unwrap();
+        assert!(!build.is_empty(), "Build should not be empty");
+        assert!(!version.is_empty(), "Build should not be empty");
+
+        println!("build: {} version: {}", build, version);
+    }
+
+    #[test]
+    fn test_write_builder_single_field() {
+        let mut query = InfluxDbQuery::write();
+
+        query.add_field("water_level", "2");
+        assert_eq!(query.build(), "measurement, water_level=2 time");
+    }
+
+    #[test]
+    fn test_write_builder_multiple_fields() {
+        let mut query = InfluxDbQuery::write();
+
+        query.add_field("water_level", "2");
+        query.add_field("boat_count", "31");
+        query.add_field("algae_content", "0.85");
+        assert_eq!(
+            query.build(),
+            "measurement, water_level=2,boat_count=31,algae_content=0.85 time"
+        );
+    }
+
+    // fixme: double space
+    // fixme: quoting / escaping of long strings
+    #[test]
+    fn test_write_builder_single_tag() {
+        let mut query = InfluxDbQuery::write();
+
+        query.add_tag("marina_manager", "Smith");
+        assert_eq!(query.build(), "measurement,marina_manager=Smith  time");
+    }
+
+    #[test]
+    fn test_write_builder_multiple_tags() {
+        let mut query = InfluxDbQuery::write();
+
+        query.add_tag("marina_manager", "Smith");
+        query.add_tag("manager_to_the_marina_manager", "Jonson");
+        assert_eq!(
+            query.build(),
+            "measurement,marina_manager=Smith,manager_to_the_marina_manager=Jonson  time"
+        );
+    }
+
+    #[test]
+    fn test_write_builder_full_query() {
+        let mut query = InfluxDbQuery::write();
+
+        query.add_field("water_level", "2");
+        query.add_field("boat_count", "31");
+        query.add_field("algae_content", "0.85");
+        query.add_tag("marina_manager", "Smith");
+        query.add_tag("manager_to_the_marina_manager", "Jonson");
+        assert_eq!(
+            query.build(),
```

Comment:
I'm particularly unhappy about this Builder pattern. Methods cannot be chained due to lifetime issues, but require some `let` variables to create a stronger binding.

Impossible: `InfluxDbQuery::write().add_field("test", "1").add_tag("my_tag", "0.85").build()`.

Any ideas for improvement?

Reply:
Well, a builder pattern in rust must consume `mut self` instead of `&self` and return `Self` or `Result<Self, Error>` or similar. The code you claimed as _impossible_ can definitely be made possible